src/chainstate/stacks/index/storage.rs

```diff
@@ -109,6 +106,67 @@ pub fn fseek_end<F: Seek>(f: &mut F) -> Result<u64, Error> {
         .map_err(Error::IOError)
 }
 
+#[derive(Clone)]
+pub struct BlockHashMap {
+    next_identifier: u32,
+    map: Vec<BlockHeaderHash>
+}
+
+trait NodeHashReader {
+    fn write_node_hash<W: Write>(&mut self, ptr: &TriePtr, w: &mut W) -> Result<(), Error>;
+}
+
+impl BlockHashMap {
```

Comment:
While I like the idea of a `BlockHashMap` in general, I don't think this implementation will fly as-is.  The reason is that it makes it all but impossible for a verifier to work without access to the prover's block map.  Is there a way the block identifiers can be deterministically generated from a _set_ of block hashes?  Or, possibly, a tree of block/parent hash pairs?  If so, then this would make it possible for a light client to independently calculate the same `BlockHashMap` as a full node, using just the block headers (which is all it has access to).  This will be necessary for the light client, since it needs the block map to walk and verify the MARF Merkle proofs.

Reply:
Right — as I replied elsewhere, the Merkle proof needs to change. Back pointers in the proof will just be the block header hash (which is what’s used in the consensus bytes for the hash as is).

Reply:
Addressed in da7f5c279d25c99ab09bb01242fa31a179878761